File: data-structures-and-algorithms/algorithm/leetCode/stack&queue/isValid/isValid.php
<?php
/**
 * @date: 2022/3/14
 * @createTime: 1:14 PM
 */

// Leetcode-20
function isValid($s)
{
    $stack = [];
    for ($i = 0; $i < strlen($s); $i++) {
        if ($s[$i] == '(' || $s[$i] == '[' || $s[$i] == '{') {
            array_push($stack, $s[$i]);
        } else {
            if (count($stack) == 0)
                return false;

            $top = array_pop($stack);

            if ($s[$i] == ')') {
                $match = '(';
            } elseif ($s[$i] == ']') {
                $match = '[';
            } else {
                $match = '{';
            }

            if ($top != $match)
                return false;
        }
    }

    if (count($stack) != 0)
        return false;

    return true;
}

File: dataStructures&algorithms/algorithm/leetCode/stack&queue/isValid/isValid.cpp
#include <iostream>
#include <cassert>
#include <string>
#include <stack>

using namespace std;

// Leetcode-20
class Solution {
public:
	bool isValid(string s) {
		stack<char> stack;
		for (int i = 0; i < s.size(); i++) {
			if (s[i] == '(' || s[i] == '[' || s[i] == '{')
				stack.push(s[i]);
			else {

				if (stack.size() == 0)
					return false;

				char c = stack.top();
				stack.pop();

				char match;
				if (s[i] == ')')
					match = '(';
				else if (s[i] == ']')
					match = '[';
				else {
					assert(s[i] == '}');
					match = '{';
				}
				
				if (c != match)
					return false;
			}
		}

		if (stack.size() != 0)
			return false;

		return true;
	}
};

int main() {

	return 0;
}